<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Importing datasets into OpenGWAS [NEW WORKFLOW] • GwasDataImport</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Importing datasets into OpenGWAS [NEW WORKFLOW]">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">GwasDataImport</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-vignettes" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Vignettes</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-vignettes">
<li><a class="dropdown-item" href="../articles/index.html">List of vignettes</a></li>
    <li><a class="dropdown-item" href="../articles/import_pipeline_new.html">Importing datasets into OpenGWAS [NEW]</a></li>
    <li><a class="dropdown-item" href="../articles/import_pipeline.html">Importing datasets into OpenGWAS</a></li>
    <li><a class="dropdown-item" href="../articles/edit.html">Edit existing GWAS meta data</a></li>
    <li><a class="dropdown-item" href="../articles/ebi_workflow.html">EBI upload pipeline</a></li>
    <li><a class="dropdown-item" href="../articles/liftover.html">Automating liftover</a></li>
    <li><a class="dropdown-item" href="../articles/testrun.html">Test run for upload system</a></li>
    <li><a class="dropdown-item" href="../articles/get_positions.html">Looking up chromosome and position for large sets of rs IDs</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Functions</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/mrcieu/GwasDataImport">Source</a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Importing datasets into OpenGWAS [NEW WORKFLOW]</h1>
            
      

      <div class="d-none name"><code>import_pipeline_new.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="as-a-contributor">As a Contributor<a class="anchor" aria-label="anchor" href="#as-a-contributor"></a>
</h2>
<p>This role enables you to contribute datasets (create metadata, upload
file for QC, check QC report and submit dataset for approval) to
OpenGWAS.</p>
<p>You may contact Admin (details TBC) and request to be added as a
Contributor. You will be granted access to <a href="https://api.opengwas.io/contribution" class="external-link uri">https://api.opengwas.io/contribution</a>.</p>
<p>You also need R/GwasDataImport installed and your OpenGWAS JWT
(token) set up in your R environment.</p>
</div>
<div class="section level2">
<h2 id="prerequisites">Prerequisites<a class="anchor" aria-label="anchor" href="#prerequisites"></a>
</h2>
<p>Each dataset will be a <code>.txt</code> or <code>.txt.gz</code>
file. The content looks like this:</p>
<pre><code>ID          ALT  REF  BETA     SE      PVALUE   AF       N       ZVALUE  INFO  CHROM  POS
rs10399878  G    A    0.0118   0.016   0.4608   0.9569   124787  NA      NA    1      1239953
rs1039063   G    T    0.0026   0.0036  0.4702   0.55     236102  NA      NA    1      2281978
rs1039100   G    A    0.0033   0.0047  0.4826   NA       221290  NA      NA    1      2286947
rs10158583  A    G    0.0099   0.0059  0.09446  0.075    321197  NA      NA    1      3144068
rs10157420  C    T    -0.0038  0.0075  0.6124   0.05     234171  NA      NA    1      3146497</code></pre>
<p>The header row is not required - you will need to specify the
mappings in later steps. The header row (if any) will be ignored by the
pipeline so it’s okay to leave it as-is.</p>
<p>These columns are required for each dataset:</p>
<ul>
<li>chromosome</li>
<li>position</li>
<li>beta</li>
<li>se</li>
<li>effect allele</li>
<li>other allele</li>
<li>pval</li>
</ul>
<p>Note: you need to remove any leading <code>0</code> from the
chromosome values, e.g. <code>07</code> -&gt; <code>7</code>.</p>
<p>Note: a row will be removed by the pipeline entirely if it has NA/Inf
value in any required columns.</p>
<p>These columns are optional:</p>
<ul>
<li>rsid</li>
<li>effect allele frequency</li>
<li>other allele frequency</li>
<li>number of cases</li>
<li>number of controls (or total sample size if continuous trait)</li>
<li>imputation z-score</li>
<li>imputation info score</li>
</ul>
</div>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<p>You can track your contributions via the web interface, but you will
need R/GwasDataImport anyway to upload the files.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">GwasDataImport</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make sure you have switched to the "public" API base URL</span></span>
<span><span class="fu">ieugwasr</span><span class="fu">::</span><span class="fu"><a href="https://mrcieu.github.io/ieugwasr/reference/select_api.html" class="external-link">select_api</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Make sure you are authenticated and have got the "contributor" or "admin" role </span></span>
<span><span class="fu">ieugwasr</span><span class="fu">::</span><span class="fu"><a href="https://mrcieu.github.io/ieugwasr/reference/user.html" class="external-link">user</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="metadata-and-dataset">Metadata and dataset<a class="anchor" aria-label="anchor" href="#metadata-and-dataset"></a>
</h2>
<p>For each dataset you will need to go through step 1 to 4. See also:
<a href="#import-in-bulk">Import in bulk</a>.</p>
<div class="section level3">
<h3 id="create-metadata-and-obtain-gwas-id-two-methods">1. Create metadata and obtain GWAS ID (two methods)<a class="anchor" aria-label="anchor" href="#create-metadata-and-obtain-gwas-id-two-methods"></a>
</h3>
<p>For each dataset to be uploaded, you can <strong>either</strong> use
the web interface <strong>or</strong> the R/GwasDataImport package to
create the metadata.</p>
<div class="section level4">
<h4 id="a--use-the-web-interface">1a. Use the web interface<a class="anchor" aria-label="anchor" href="#a--use-the-web-interface"></a>
</h4>
<p><a href="https://api.opengwas.io/contribution/" class="external-link uri">https://api.opengwas.io/contribution/</a> provides a webform
with dropdowns and tooltips for each field. This is handy when you are
new to this process and/or only have a few datasets to upload.</p>
<p>GWAS ID will be available on the web interface for each dataset.
Assume it’s <code>ieu-b-5137</code>.</p>
<p>In R, specify the path (assume it’s <code>~/bmi_test.txt.gz</code>)
and GWAS ID, and then mark the metadata as uploaded.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/Dataset.html">Dataset</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>filename<span class="op">=</span><span class="st">"~/bmi_test.txt.gz"</span>, igd_id<span class="op">=</span><span class="st">'ieu-b-5137'</span><span class="op">)</span></span>
<span></span>
<span><span class="va">x</span><span class="op">$</span><span class="va">metadata_uploaded</span> <span class="op">&lt;-</span> <span class="cn">TRUE</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="b--use-rgwasdataimport">1b. Use R/GwasDataImport<a class="anchor" aria-label="anchor" href="#b--use-rgwasdataimport"></a>
</h4>
<p><strong>Alternatively</strong>, you can opt to use R if you have an
array of candidate datasets since it’s easier to upload metadata of
multiple datasets in a programmable way. We recommend that you try the
web interface in 1a first before you go down this more advanced
route.</p>
<p>Assume the full path to the file is
<code>~/bmi_test.txt.gz</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Don't provide the GWAS ID</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="va"><a href="../reference/Dataset.html">Dataset</a></span><span class="op">$</span><span class="fu">new</span><span class="op">(</span>filename<span class="op">=</span><span class="st">"~/bmi_test.txt.gz"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">x</span><span class="op">$</span><span class="fu">collect_metadata</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    trait<span class="op">=</span><span class="st">"TEST - DO NOT USE 2"</span>,</span>
<span>    group_name<span class="op">=</span><span class="st">"public"</span>,</span>
<span>    build<span class="op">=</span><span class="st">"HG19/GRCh37"</span>,</span>
<span>    category<span class="op">=</span><span class="st">"Risk factor"</span>,</span>
<span>    subcategory<span class="op">=</span><span class="st">"Anthropometric"</span>,</span>
<span>    ontology<span class="op">=</span><span class="st">"NA"</span>,</span>
<span>    population<span class="op">=</span><span class="st">"Mixed"</span>,</span>
<span>    sex<span class="op">=</span><span class="st">"Males and Females"</span>,</span>
<span>    sample_size<span class="op">=</span><span class="fl">339224</span>,</span>
<span>    author<span class="op">=</span><span class="st">"Mendel GJ"</span>,</span>
<span>    year<span class="op">=</span><span class="fl">2022</span>,</span>
<span>    unit<span class="op">=</span><span class="st">"SD"</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">x</span><span class="op">$</span><span class="fu">api_metadata_upload</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>GWAS ID will be returned by the last command. Assume it’s
<code>ieu-b-5137</code>. At the same time a new record will show up on
<a href="https://api.opengwas.io/contribution/" class="external-link uri">https://api.opengwas.io/contribution/</a>.</p>
</div>
</div>
<div class="section level3">
<h3 id="modify-the-metadata-only-when-necessary">2. Modify the metadata (only when necessary)<a class="anchor" aria-label="anchor" href="#modify-the-metadata-only-when-necessary"></a>
</h3>
<p>You can modify the metadata either via the web interface
(recommended) or through R/GwasDataImport, regardless of how you created
the metadata (i.e. metadata created via the R package can be modified on
the web interface, or vice versa).</p>
<p>Note that the metadata can only be modified when there is no QC
pipeline associated, because at the last step when generating the QC
report, the metadata will be hardcoded in the report. Metadata can only
be modified when</p>
<ul>
<li>(a) the file has not been uploaded for QC, or</li>
<li>(b) the QC pipeline had already finished and you decided to “delete
uploaded files and QC products but keep the metadata”, which effectly
reverted the state to (a)</li>
</ul>
</div>
<div class="section level3">
<h3 id="format-the-file-and-upload-for-qc-rgwasdataimport">3. Format the file and upload for QC (R/GwasDataImport)<a class="anchor" aria-label="anchor" href="#format-the-file-and-upload-for-qc-rgwasdataimport"></a>
</h3>
<p>Always check that the GWAS ID and path information stored are
accurate.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="va">igd_id</span></span>
<span><span class="va">x</span><span class="op">$</span><span class="va">filename</span></span></code></pre></div>
<p>Specify column mapping (1-indexed) (check <a href="https://mrcieu.github.io/GwasDataImport/reference/Dataset.html#method-Dataset-determine_columns">docs</a>
for parameter names):</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">determine_columns</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  chr_col<span class="op">=</span><span class="fl">11</span>,</span>
<span>  pos_col<span class="op">=</span><span class="fl">12</span>,</span>
<span>  ea_col<span class="op">=</span><span class="fl">2</span>,</span>
<span>  oa_col<span class="op">=</span><span class="fl">3</span>,</span>
<span>  beta_col<span class="op">=</span><span class="fl">4</span>,</span>
<span>  se_col<span class="op">=</span><span class="fl">5</span>,</span>
<span>  pval_col<span class="op">=</span><span class="fl">6</span>,</span>
<span>  snp_col<span class="op">=</span><span class="fl">1</span>,</span>
<span>  eaf_col<span class="op">=</span><span class="fl">7</span>,</span>
<span>  ncontrol_col<span class="op">=</span><span class="fl">8</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Format the dataset (this may take a while):</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">format_dataset</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>Use the output to double-check the mapping. If necessary, run
<code>x$determine_columns(...)</code> again with the correct
mapping.</p>
<p>Upload the dataset (this may take a while):</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">api_gwasdata_upload</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>You will see the “Dataset has been added to the pipeline” message if
the upload was successful.</p>
<p>And finally don’t forget to clean up the working directory:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span><span class="op">$</span><span class="fu">delete_wd</span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="check-qc-pipeline-state-and-report-and-submit-for-approval">4. Check QC pipeline state and report, and submit for approval<a class="anchor" aria-label="anchor" href="#check-qc-pipeline-state-and-report-and-submit-for-approval"></a>
</h3>
<p>On <a href="https://api.opengwas.io/contribution" class="external-link uri">https://api.opengwas.io/contribution</a> you can click the
<strong>2. QC</strong> tab of the dataset popup and check pipeline
state.</p>
<p>For each dataset, you should review the QC report when it’s available
and decide whether to submit the dataset for approval or not. You will
have the following options:</p>
<ul>
<li>Submit the dataset for approval
<ul>
<li>Go to the <strong>3. Approval &amp; Release</strong> tab and
submit</li>
</ul>
</li>
<li>Re-upload the file for QC
<ul>
<li>Go to the <strong>1. Metadata</strong> tab and “delete uploaded
files and QC products”</li>
</ul>
</li>
<li>Discard the dataset
<ul>
<li>Go to the <strong>1. Metadata</strong> tab and “delete
everything”</li>
</ul>
</li>
</ul>
<p>You may also use the checkboxes on the main screen to select datasets
and submit for approval in bulk.</p>
</div>
</div>
<div class="section level2">
<h2 id="import-in-bulk">Import in bulk<a class="anchor" aria-label="anchor" href="#import-in-bulk"></a>
</h2>
<p>If you have multiple datasets you may want to write an R snippet to
semi-automate this process.</p>
<p><a href="#setup">Set up</a> for only once, then for each dataset go
through step 1b, 2 and 3. Finally visit the portal in step 4 and use the
checkboxes to submit in bulk.</p>
</div>
<div class="section level2">
<h2 id="whats-next">What’s next<a class="anchor" aria-label="anchor" href="#whats-next"></a>
</h2>
<p>Admins will review and approve/reject each datasets.</p>
</div>
<div class="section level2">
<h2 id="known-issues">Known issues<a class="anchor" aria-label="anchor" href="#known-issues"></a>
</h2>
<ul>
<li>GwasDataImport has the ability to liftover, but we recommend you to
convert all records to <strong>GRCh37</strong> if possible.</li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Gibran Hemani, Philip Haycock, Tom Palmer.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
